src/jvmMain/kotlin/com/sdercolin/vlabeler/video/VideoPlayer.kt

```diff
@@ -0,0 +1,111 @@
+// see https://github.com/JetBrains/compose-jb/pull/1738 for direct control of awt/swing
+@file:Suppress("INVISIBLE_REFERENCE", "INVISIBLE_MEMBER")
+
+package com.sdercolin.vlabeler.video
+
+import androidx.compose.foundation.layout.Box
+import androidx.compose.runtime.Composable
+import androidx.compose.runtime.DisposableEffect
+import androidx.compose.runtime.SideEffect
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.awt.LocalLayerContainer
+import androidx.compose.ui.geometry.Offset
+import androidx.compose.ui.layout.onGloballyPositioned
+import androidx.compose.ui.platform.LocalDensity
+import androidx.compose.ui.unit.round
+import com.sdercolin.vlabeler.env.Log
+import com.sdercolin.vlabeler.env.isMacOS
+import uk.co.caprica.vlcj.factory.discovery.NativeDiscovery
+import uk.co.caprica.vlcj.player.base.MediaPlayer
+import uk.co.caprica.vlcj.player.component.CallbackMediaPlayerComponent
+import uk.co.caprica.vlcj.player.component.EmbeddedMediaPlayerComponent
+import uk.co.caprica.vlcj.player.component.InputEvents
+import java.awt.BorderLayout
+import javax.swing.JPanel
+
+class MiniVideo() {
+    var mediaPlayerComponent: Any? = null
+    lateinit var mediaPlayer: MediaPlayer
+    var lastTime: Long? = null
+        private set
+    lateinit var toggleCallback: (on: Boolean) -> Unit
+
+    fun init(toggleCallback: (on: Boolean) -> Unit) {
+        Log.info("VideoPlayer init")
+
+        NativeDiscovery().discover()
+        // see https://github.com/caprica/vlcj/issues/887#issuecomment-503288294 for why we're using CallbackMediaPlayerComponent for macOS.
+        mediaPlayerComponent = when {
+            isMacOS -> CallbackMediaPlayerComponent(null, null, InputEvents.NONE, true, null)
+            else -> EmbeddedMediaPlayerComponent(null, null, null, InputEvents.NONE, null)
+        }
+        mediaPlayer = getMediaPlayer(mediaPlayerComponent!!)
+
+        this.toggleCallback = toggleCallback
+    }
+
+    /**
+     * To return mediaPlayer from player components.
+     * The method names are same, but they don't share the same parent/interface.
+     * That's why need this method.
+     */
+    private fun getMediaPlayer(component: Any): MediaPlayer =
```

Comment:
If you implement an interface as I commented above we will not need this.

Reply:
To Minimize redundant type casting code, we can keep the type information in a `run` block:
```
        mediaPlayerComponent = when {
            isMacOS -> {
                CallbackMediaPlayerComponent(null, null, InputEvents.NONE, true, null)
                    .run{
                        mediaPlayer = mediaPlayer()
                        MediaPlayerComponent(this)
                    }
            }
            else -> {
                EmbeddedMediaPlayerComponent(null, null, null, InputEvents.NONE, null)
                    .run{
                        mediaPlayer = mediaPlayer()
                        MediaPlayerComponent(this)
                    }
            }
        }
```
There are still some repeated code, but are necessary.